File: windows/src/Nabira.Win/Core/TypoCorrector.cs
namespace Nabira.Win.Core;

internal static class TypoCorrector
{
    private static readonly Dictionary<string, string> Overrides = new(StringComparer.OrdinalIgnoreCase)
    {
        ["adress"] = "address", ["becuase"] = "because", ["begining"] = "beginning",
        ["bokk"] = "book", ["comming"] = "coming", ["definately"] = "definitely",
        ["freind"] = "friend", ["goverment"] = "government", ["helo"] = "hello",
        ["langauge"] = "language", ["occured"] = "occurred", ["recieve"] = "receive",
        ["seperate"] = "separate", ["teh"] = "the", ["thier"] = "their",
        ["tomorow"] = "tomorrow", ["untill"] = "until", ["watre"] = "water",
        ["wich"] = "which", ["wierd"] = "weird",
        ["агенство"] = "агентство", ["будующее"] = "будущее", ["здраствуйте"] = "здравствуйте",
        ["извените"] = "извините", ["интиресный"] = "интересный", ["карова"] = "корова",
        ["коментарий"] = "комментарий", ["координально"] = "кардинально",
        ["ошыбка"] = "ошибка", ["пажалуйста"] = "пожалуйста", ["превет"] = "привет",
        ["програма"] = "программа", ["професор"] = "профессор", ["работаеть"] = "работает",
        ["рассказатьь"] = "рассказать", ["сабака"] = "собака", ["сделанно"] = "сделано",
        ["учавствовать"] = "участвовать"
    };

    public static string? Replacement(string input, string language)
    {
        if (!Eligible(input, language)) return null;
        if (Overrides.TryGetValue(input, out string? known)) return PreserveCase(input, known);
        if (WordFrequency.IsKnown(input, language)) return null;
        if (Dict.Available && Dict.IsValidWord(input.ToLowerInvariant(), language)) return null;

        string source = input.ToLowerInvariant();
        var suggestions = new List<string>();
        if (Dict.Available) suggestions.AddRange(Dict.Suggestions(source, language));
        foreach (string suggestion in SymmetricDeleteSpeller.Suggestions(source, language))
            if (!suggestions.Contains(suggestion, StringComparer.OrdinalIgnoreCase))
                suggestions.Add(suggestion);
        if (CanonicalCaseSuggestion(input, suggestions, language) is { } canonicalCase)
            return canonicalCase;

        int maxDistance = source.Length >= 7 ? 2 : 1;
        var candidates = suggestions
            .Select(s => s.ToLowerInvariant())
            .Where(s => s != source && s.All(char.IsLetter))
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .Select((word, index) => new
            {
                Word = word,
                Index = index,
                Frequency = WordFrequency.Frequency(word, language) ?? 0,
                Distance = Damerau(source, word)
            })
            .Where(value => value.Frequency >= WordFrequency.MinimumKnownFrequency
                && value.Distance > 0 && value.Distance <= maxDistance)
            .Select(value => new
            {
                value.Word,
                value.Frequency,
                Score = Math.Log10(value.Frequency + 1) * 1.4
                    - value.Distance * 4.0
                    + (value.Index == 0 ? 2.5 : Math.Max(0, 1.2 - value.Index * 0.2))
                    + (ConsonantSkeleton(source, language) == ConsonantSkeleton(value.Word, language) ? 2.4 : 0)
                    + (RestoresMissingRepeatedLetter(source, value.Word) ? 4.1 : 0)
                    + (source[0] == value.Word[0] ? 2.0 : 0)
                    + (source[^1] == value.Word[^1] ? 1.0 : 0)
            })
            .OrderByDescending(value => value.Score)
            .ThenByDescending(value => value.Frequency)
            .ToList();
        if (candidates.Count == 0) return null;
        if (candidates.Count > 1 && candidates[0].Score - candidates[1].Score < 0.35) return null;
        return PreserveCase(input, candidates[0].Word);
    }

    /// <summary>Preserves a system-provided mixed-case product spelling such as iPhone,
    /// macOS or GitHub. Initial-capital-only names remain context-dependent and are ignored.</summary>
    internal static string? CanonicalCaseSuggestion(
        string input,
        IEnumerable<string> suggestions,
        string language)
    {
        foreach (string suggestion in suggestions)
        {
            if (suggestion == input ||
                !suggestion.Equals(input, StringComparison.OrdinalIgnoreCase) ||
                !suggestion.Skip(1).Any(char.IsUpper) ||
                !suggestion.All(char.IsLetter)) continue;

            bool cyrillic = suggestion.All(c => c is >= '\u0400' and <= '\u052F');
            bool latin = suggestion.All(c => c is >= 'A' and <= 'Z' or >= 'a' and <= 'z');
            if (language == "ru" ? cyrillic : language == "en" && latin)
                return suggestion;
        }
        return null;
    }

    private static bool Eligible(string word, string language)
    {
        if (word.Length is < 3 or > 30 || !word.All(char.IsLetter)) return false;
        if (word.Skip(1).Any(char.IsUpper)) return false;
        bool cyr = word.All(c => c is >= '\u0400' and <= '\u052F');
        bool latin = word.All(c => c is >= 'A' and <= 'Z' or >= 'a' and <= 'z');
        return language == "ru" ? cyr : language == "en" && latin;
    }

    private static string PreserveCase(string source, string target) =>
        char.IsUpper(source[0]) ? char.ToUpperInvariant(target[0]) + target[1..] : target;

    internal static int Damerau(string lhs, string rhs)
    {
        int[,] d = new int[lhs.Length + 1, rhs.Length + 1];
        for (int i = 0; i <= lhs.Length; i++) d[i, 0] = i;
        for (int j = 0; j <= rhs.Length; j++) d[0, j] = j;
        for (int i = 1; i <= lhs.Length; i++)
        for (int j = 1; j <= rhs.Length; j++)
        {
            int cost = lhs[i - 1] == rhs[j - 1] ? 0 : 1;
            d[i, j] = Math.Min(Math.Min(d[i - 1, j] + 1, d[i, j - 1] + 1), d[i - 1, j - 1] + cost);
            if (i > 1 && j > 1 && lhs[i - 1] == rhs[j - 2] && lhs[i - 2] == rhs[j - 1])
                d[i, j] = Math.Min(d[i, j], d[i - 2, j - 2] + cost);
        }
        return d[lhs.Length, rhs.Length];
    }

    internal static void WarmUp() => SymmetricDeleteSpeller.WarmUp();

    private static string ConsonantSkeleton(string word, string language)
    {
        const string RussianVowels = "аеёиоуыэюя";
        const string EnglishVowels = "aeiouy";
        string vowels = language.StartsWith("ru", StringComparison.OrdinalIgnoreCase)
            ? RussianVowels : EnglishVowels;
        return new string(word.Where(c => !vowels.Contains(c)).ToArray());
    }

    private static bool RestoresMissingRepeatedLetter(string source, string candidate)
    {
        if (candidate.Length != source.Length + 1) return false;
        for (int index = 1; index < candidate.Length; index++)
            if (candidate[index] == candidate[index - 1] && candidate.Remove(index, 1) == source)
                return true;
        return false;
    }
}

File: macos/Sources/Nabira/TypoCorrector.swift
import AppKit
import Foundation

/// Conservative offline typo correction for Russian and English.
///
/// AppleSpell supplies language-aware candidates.  A bundled frequency table and
/// structural typo signals rank those candidates so we do not blindly accept the
/// first suggestion (for example, `adress` must become `address`, not `dress`).
enum TypoCorrector {
    private struct Candidate {
        let word: String
        let score: Double
        let isSystemCorrection: Bool
        let structuralMatch: Bool
    }

    @MainActor private static let checker = NSSpellChecker.shared
    private static let supportedLanguages: Set<String> = ["en", "ru"]
    private static let maxCandidateCount = 8

    /// Frequent, unambiguous errors where a generic spell checker is known to prefer
    /// the wrong grammatical form (`коментарий` → `комментарии`) or another valid word.
    /// Keep this list deliberately small; the candidate model handles the long tail.
    private static let highConfidenceOverrides: [String: [String: String]] = [
        "en": [
            "adress": "address", "becuase": "because", "begining": "beginning",
            "bokk": "book", "comming": "coming", "definately": "definitely",
            "freind": "friend", "goverment": "government", "helo": "hello",
            "langauge": "language", "occured": "occurred", "recieve": "receive",
            "seperate": "separate", "teh": "the", "thier": "their",
            "tomorow": "tomorrow", "untill": "until", "watre": "water",
            "wich": "which", "wierd": "weird",
        ],
        "ru": [
            "агенство": "агентство", "будующее": "будущее", "здраствуйте": "здравствуйте",
            "извените": "извините", "интиресный": "интересный", "карова": "корова",
            "коментарий": "комментарий", "координально": "кардинально",
            "ошыбка": "ошибка", "пажалуйста": "пожалуйста", "превет": "привет",
            "програма": "программа", "професор": "профессор", "работаеть": "работает",
            "рассказатьь": "рассказать", "сабака": "собака", "сделанно": "сделано",
            "учавствовать": "участвовать",
        ],
    ]

    static func warmUp() {
        WordFrequency.warmUp()
    }

    /// Returns a confident correction while preserving the capitalization of the input.
    /// Unknown words, identifiers, acronyms and ambiguous candidates are left untouched.
    @MainActor
    static func replacement(for input: String, language: String, context: String? = nil) -> String? {
        let normalized = input.precomposedStringWithCanonicalMapping
        let lang = String(language.lowercased().prefix(2))
        guard supportedLanguages.contains(lang),
              isEligible(normalized, language: lang),
              Dict.isAvailable(lang),
              !Dict.isValidWord(normalized.lowercased(), lang: lang),
              let lexicon = WordFrequency.table(language: lang) else { return nil }

        if let replacement = highConfidenceOverrides[lang]?[normalized.lowercased()] {
            return preserveCapitalization(of: normalized, in: replacement)
        }
        // The bundled frequency corpus contains common brands and product names that
        // AppleSpell may not accept (for example `iphone`).  Treat an established
        // corpus word as intentional instead of replacing it with a nearby common word
        // such as `phone`. Explicit high-confidence typo overrides above still win.
        if (lexicon[normalized.lowercased()] ?? 0) >= 20 { return nil }
        // Outside the explicit high-confidence list, a capitalized unknown token is
        // more likely to be a person's name or a brand than a typo.
        guard normalized.first?.isUppercase != true else { return nil }

        let query = spellQuery(word: normalized, context: context)
        let range = query.range
        let text = query.text
        let correction = checker.correction(
            forWordRange: range,
            in: text,
            language: lang,
            inSpellDocumentWithTag: 0
        )
        let guesses = checker.guesses(
            forWordRange: range,
            in: text,
            language: lang,
            inSpellDocumentWithTag: 0
        ) ?? []

        guard let selected = selectCandidate(
            typed: normalized,
            correction: correction,
            guesses: Array(guesses.prefix(maxCandidateCount)),
            frequencies: lexicon,
            language: lang
        ) else { return nil }
        return preserveCapitalization(of: normalized, in: selected)
    }

    /// Detects the correction language from the actual script, not the active layout.
    static func language(for word: String) -> String? {
        var hasLatin = false
        var hasCyrillic = false
        for scalar in word.unicodeScalars {
            switch scalar.value {
            case 0x41...0x5A, 0x61...0x7A:
                hasLatin = true
            case 0x0400...0x04FF:
                hasCyrillic = true
            default:
                return nil
            }
        }
        if hasLatin == hasCyrillic { return nil }
        return hasCyrillic ? "ru" : "en"
    }

    /// Pure candidate ranking, kept internal for deterministic unit tests.
    static func selectCandidate(
        typed: String,
        correction: String?,
        guesses: [String],
        frequencies: [String: Int],
        language: String
    ) -> String? {
        // AppleSpell often returns the canonical spelling of a product only by case
        // (`iphone` -> `iPhone`). Accept that exact-letter correction before lowercasing
        // candidates; otherwise it disappears as a no-op and the next guess (`phone`)
        // can be selected as a false typo correction.
        if let canonicalCase = canonicalCaseCorrection(
            typed: typed,
            correction: correction,
            language: language
        ) {
            return canonicalCase
        }

        let source = typed.lowercased()
        let sourceWasLowercase = typed == typed.lowercased()
        let systemCorrection = normalizedCandidate(
            correction,
            source: source,
            language: language,
            rejectCapitalized: sourceWasLowercase
        )
        var ordered: [String] = []
        if let systemCorrection { ordered.append(systemCorrection) }
        for guess in guesses {
            guard let word = normalizedCandidate(
                guess,
                source: source,
                language: language,
                rejectCapitalized: sourceWasLowercase
            ),
                  !ordered.contains(word) else { continue }
            ordered.append(word)
        }

        let maxDistance = source.count >= 7 ? 2 : 1
        let sourceSkeleton = consonantSkeleton(source, language: language)
        var ranked: [Candidate] = []
        for word in ordered {
            guard let frequency = frequencies[word], frequency >= 20 else { continue }
            let distance = damerauLevenshteinDistance(source, word)
            guard distance > 0, distance <= maxDistance else { continue }

            let guessIndex = guesses.firstIndex { $0.lowercased() == word }
            let isCorrection = word == systemCorrection
            let sameSkeleton = !sourceSkeleton.isEmpty
                && sourceSkeleton == consonantSkeleton(word, language: language)
            let restoresRepeatedLetter = restoresMissingRepeatedLetter(source, word)
            let structuralMatch = sameSkeleton || restoresRepeatedLetter
            let prefix = commonPrefixLength(source, word)

            var score = log10(Double(frequency) + 1) * 1.4
            score -= Double(distance) * 4.0
            if isCorrection { score += 2.5 }
            if let guessIndex { score += max(0.0, 1.2 - Double(guessIndex) * 0.2) }
            if sameSkeleton { score += 2.4 }
            if restoresRepeatedLetter { score += 4.1 }
            score += Double(prefix) * 0.15
            if source.first == word.first { score += 2.0 }
            if source.last == word.last { score += 1.0 }
            if source.count == word.count { score += 0.25 }

            ranked.append(Candidate(
                word: word,
                score: score,
                isSystemCorrection: isCorrection,
                structuralMatch: structuralMatch
            ))
        }

        ranked.sort { lhs, rhs in
            if lhs.score != rhs.score { return lhs.score > rhs.score }
            return (frequencies[lhs.word] ?? 0) > (frequencies[rhs.word] ?? 0)
        }
        guard let best = ranked.first else { return nil }
        let margin = ranked.count > 1 ? best.score - ranked[1].score : .infinity

        // AppleSpell's automatic correction is already conservative.  A structural
        // match can override it only with a clear lead (adress/address, helo/hello).
        if best.isSystemCorrection {
            guard margin >= 0.35 || ranked.count == 1 else { return nil }
        } else {
            guard best.structuralMatch, margin >= 0.75 || ranked.count == 1 else { return nil }
        }
        return best.word
    }

    /// A conservative case-only canonicalization for mixed-case product names.
    /// Initial-capital-only suggestions (`john` -> `John`) remain untouched because
    /// they are too context-dependent for an automatic correction.
    static func canonicalCaseCorrection(
        typed: String,
        correction: String?,
        language: String
    ) -> String? {
        guard let correction = correction?.precomposedStringWithCanonicalMapping,
              correction != typed,
              correction.caseInsensitiveCompare(typed) == .orderedSame,
              correction.dropFirst().contains(where: { $0.isUppercase }),
              correction.allSatisfy({ $0.isLetter }),
              self.language(for: correction) == String(language.lowercased().prefix(2)) else {
            return nil
        }
        return correction
    }

    static func damerauLevenshteinDistance(_ lhs: String, _ rhs: String) -> Int {
        let a = Array(lhs)
        let b = Array(rhs)
        guard !a.isEmpty else { return b.count }
        guard !b.isEmpty else { return a.count }

        var previousPrevious = Array(0...b.count)
        var previous = previousPrevious
        for i in 1...a.count {
            var current = Array(repeating: 0, count: b.count + 1)
            current[0] = i
            for j in 1...b.count {
                let substitution = previous[j - 1] + (a[i - 1] == b[j - 1] ? 0 : 1)
                current[j] = min(previous[j] + 1, current[j - 1] + 1, substitution)
                if i > 1, j > 1, a[i - 1] == b[j - 2], a[i - 2] == b[j - 1] {
                    current[j] = min(current[j], previousPrevious[j - 2] + 1)
                }
            }
            previousPrevious = previous
            previous = current
        }
        return previous[b.count]
    }

    private static func isEligible(_ word: String, language: String) -> Bool {
        guard word.count >= 3, word.count <= 30,
              word.allSatisfy({ $0.isLetter }),
              self.language(for: word) == language,
              !LayoutDetector.isAllCaps(word),
              !LayoutDetector.looksLikeCodeIdentifier(word) else { return false }

        // Preserve normal sentence capitalization, but skip unusual inner capitals.
        let uppercasePositions = word.enumerated().compactMap { $0.element.isUppercase ? $0.offset : nil }
        return uppercasePositions.isEmpty || uppercasePositions == [0]
    }

    private static func normalizedCandidate(_ raw: String?, source: String, language: String,
                                            rejectCapitalized: Bool) -> String? {
        guard let raw else { return nil }
        if rejectCapitalized, raw.first?.isUppercase == true { return nil }
        let word = raw.precomposedStringWithCanonicalMapping.lowercased()
        guard word != source,
              word.allSatisfy({ $0.isLetter }),
              self.language(for: word) == String(language.prefix(2)) else { return nil }
        return word
    }

    private static func spellQuery(word: String, context: String?) -> (text: String, range: NSRange) {
        guard let context,
              context.count <= 160,
              let swiftRange = context.range(of: word, options: [.backwards, .caseInsensitive]) else {
            return (word, NSRange(location: 0, length: (word as NSString).length))
        }
        return (context, NSRange(swiftRange, in: context))
    }

    private static func preserveCapitalization(of source: String, in lowerTarget: String) -> String {
        // A mixed-case system correction (`iPhone`, `macOS`, `GitHub`) already carries
        // its canonical capitalization and must not be flattened or title-cased.
        if lowerTarget.dropFirst().contains(where: { $0.isUppercase }) { return lowerTarget }
        guard source.first?.isUppercase == true else { return lowerTarget }
        guard let first = lowerTarget.first else { return lowerTarget }
        return String(first).uppercased() + lowerTarget.dropFirst()
    }

    private static func consonantSkeleton(_ word: String, language: String) -> String {
        let vowels: Set<Character> = language == "ru"
            ? Set("аеёиоуыэюя")
            : Set("aeiouy")
        return String(word.filter { !vowels.contains($0) })
    }

    private static func restoresMissingRepeatedLetter(_ source: String, _ candidate: String) -> Bool {
        guard candidate.count == source.count + 1 else { return false }
        let chars = Array(candidate)
        for index in 1..<chars.count where chars[index] == chars[index - 1] {
            var reduced = chars
            reduced.remove(at: index)
            if String(reduced) == source { return true }
        }
        return false
    }

    private static func commonPrefixLength(_ lhs: String, _ rhs: String) -> Int {
        zip(lhs, rhs).prefix { pair in pair.0 == pair.1 }.count
    }

}
